port/linux/omrvmem.c

```diff
@@ -803,46 +811,48 @@ omrvmem_reserve_memory_ex(struct OMRPortLibrary *portLibrary, struct J9PortVmemI
+#endif /* MAP_HUGETLB */
+		alignmentInBytes = OMR_MAX(params->pageSize, params->alignmentInBytes);
+		minimumGranule = OMR_MIN(params->pageSize, params->alignmentInBytes);
 
 		/* Make sure that the alignment is a multiple of both requested alignment and page size (enforces that arguments are powers of two and, thus, their max is their lowest common multiple) */
-		if ((0 == largePageMinimumGranule) || (0 == (largePageAlignmentInBytes % largePageMinimumGranule))) {
-			memoryPointer = reserveLargePages(portLibrary, identifier, category, params->byteAmount, params->startAddress, params->endAddress, params->pageSize, largePageAlignmentInBytes, params->options, params->mode);
-		}
-		if (NULL == memoryPointer) {
-			/* If strict page size flag is not set try again with default page size */
-			if (OMR_ARE_NO_BITS_SET(params->options, OMRPORT_VMEM_STRICT_PAGE_SIZE)) {
-#if defined(OMRVMEM_DEBUG)
-				printf("\t\t\tNULL == memoryPointer, reverting to default pages\n");
-				fflush(stdout);
-#endif
-				uintptr_t defaultPageSize = PPG_vmem_pageSize[0];
-				uintptr_t alignmentInBytes = OMR_MAX(defaultPageSize, params->alignmentInBytes);
-				uintptr_t minimumGranule = OMR_MIN(defaultPageSize, params->alignmentInBytes);
-
-				/* Make sure that the alignment is a multiple of both requested alignment and page size (enforces that arguments are powers of two and, thus, their max is their lowest common multiple) */
-				if ((0 == minimumGranule) || (0 == (alignmentInBytes % minimumGranule))) {
-					memoryPointer = getMemoryInRangeForDefaultPages(portLibrary, identifier, category, params->byteAmount, params->startAddress, params->endAddress, alignmentInBytes, params->options, params->mode);
-				}
+		if ((0 == minimumGranule) || (0 == (alignmentInBytes % minimumGranule))) {
+			if (0 != (params->mode & OMRPORT_VMEM_MEMORY_MODE_HUGE_PAGES)) {
+				memoryPointer = getMemoryInRange(portLibrary, identifier, category, params->byteAmount, params->startAddress, params->endAddress, alignmentInBytes, params->options, params->mode);
 			} else {
-				update_vmemIdentifier(identifier, NULL, NULL, 0, 0, 0, 0, 0, NULL, -1);
+				memoryPointer = reserveLargePages(portLibrary, identifier, category, params->byteAmount, params->startAddress, params->endAddress, params->pageSize, alignmentInBytes, params->options, params->mode);
 			}
 		}
-	} else {
-		/* If the pageSize is not one of the supported page sizes, error */
+
+		if (NULL == memoryPointer && (OMR_ARE_NO_BITS_SET(params->options, OMRPORT_VMEM_STRICT_PAGE_SIZE))) {
+			/* Get rid of large pages and retry with default pages */
+			params->mode &= ~OMRPORT_VMEM_MEMORY_MODE_HUGE_PAGES;
+		} else {
+			goto exit;
+		}
+	/* If the pageSize is not one of the supported page sizes, error */
+	} else if (PPG_vmem_pageSize[0] != params->pageSize) {
 		update_vmemIdentifier(identifier, NULL, NULL, 0, 0, 0, 0, 0, NULL, -1);
 		Trc_PRT_vmem_omrvmem_reserve_memory_unsupported_page_size(params->pageSize);
+		goto exit;
+	}
```

Comment:
I would say bad style here. You should not use assumption that there are only two elements in this array (element 1 was handled before and if there is not an element 0 so error - here and now it happen to be correct for Linux). Also (I am not sure how accurate it might be so not strong on this) I would avoid to use `goto` in this method but try to do refactoring it using if/then/else.

Reply:
I thought the original code was making this same assumption. In the original code there was something similar, I just swapped place of PPG_vmem_pageSize[0] with PPG_vmem_pageSize[1]. Where in the original code it first checks if pageSize is 0 (that didn't change). Then, in the original code it checks PPG_vmem_pageSize[0] first to then check PPG_vmem_pageSize[1] and if none of them matches it throws an error. Maybe I missed something? In my changes I first check for PPG_vmem_pageSize[1] to then check for PPG_vmem_pageSize[0], also throwing an error if none of them are met. I do agree with the goto, I thought it would be cleaner doing this way. I'll try doing it with if/then/else

Reply:
Sorry, I was not clear. Your code is not wrong functionally. However:
this is old code structure and it is clear because all cases are explicit:
```
if (PPG_vmem_pageSize[0] == pageSize) {
...
} else if (PPG_vmem_pageSize[1] == pageSize) {
...
} else {
    /* error */
}
```
In your version:
```
if (PPG_vmem_pageSize[1] == pageSize) {
...
} else if (PPG_vmem_pageSize[0] != pageSize) {
    /* error */
} else {
    /* What is left? Small pages only I presume... */
}
```
it is not so clear already. And with number of `goto`s it is even less readable.

So please try to have code to be structured better. And any time if code does not look easy to read it might be a sign that code might be structured better... and be easier to debug now and service in future.

Reply:
Yeah I agree, it ended up being compact but not as readable, will update it